packages/queued-request-controller/src/QueuedRequestController.ts

```diff
@@ -0,0 +1,136 @@
+import type { RestrictedControllerMessenger } from '@metamask/base-controller';
+import { BaseControllerV2 } from '@metamask/base-controller';
+import type { Patch } from 'immer';
+
+const controllerName = 'QueuedRequestController';
+
+// should serialize and persist these values.
+const stateMetadata = {
+  queue: { persist: false, anonymous: false },
+};
+
+const getDefaultState = () => ({});
+
+export const QueuedRequestControllerActionTypes = {
+  enqueueRequest: `${controllerName}:enqueueRequest` as const,
+};
+
+export const QueuedRequestControllerEventTypes = {
+  stateChange: `${controllerName}:stateChange` as const,
+  countChanged: `${controllerName}:countChanged` as const,
+};
+
+export type QueuedRequestControllerState = {
+  [k: string]: any;
+};
+
+export type QueuedRequestControllerStateChangeEvent = {
+  type: typeof QueuedRequestControllerEventTypes.stateChange;
+  payload: [QueuedRequestControllerState, Patch[]];
+};
+
+export type QueuedRequestControllerCountChangedEvent = {
+  type: typeof QueuedRequestControllerEventTypes.countChanged;
+  payload: [number];
+};
+
+export type QueuedRequestControllerEnqueueRequestAction = {
+  type: typeof QueuedRequestControllerActionTypes.enqueueRequest;
+  handler: QueuedRequestController['enqueueRequest'];
+};
+
+export type QueuedRequestControllerEvents =
+  | QueuedRequestControllerStateChangeEvent
+  | QueuedRequestControllerCountChangedEvent;
+
+export type QueuedRequestControllerActions =
+  QueuedRequestControllerEnqueueRequestAction;
+
+export type QueuedRequestControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  QueuedRequestControllerActions,
+  QueuedRequestControllerEvents,
+  string,
+  string
+>;
+
+export type QueuedRequestControllerOptions = {
+  messenger: QueuedRequestControllerMessenger;
+};
+
+/**
+ * Controller for requesting encryption public key requests requiring user approval.
+ */
+export class QueuedRequestController extends BaseControllerV2<
+  typeof controllerName,
+  QueuedRequestControllerState,
+  QueuedRequestControllerMessenger
+> {
+  private currentRequest: Promise<unknown> = Promise.resolve();
+
+  private count = 0;
+
+  /**
+   * Construct a EncryptionPublicKey controller.
+   *
+   * @param options - The controller options.
+   * @param options.messenger - The restricted controller messenger for the QueuedRequestController
+   */
+  constructor({ messenger }: QueuedRequestControllerOptions) {
+    super({
+      name: controllerName,
+      metadata: stateMetadata,
+      messenger,
+      state: getDefaultState(),
+    });
+    this.#registerMessageHandlers();
+  }
+
+  #registerMessageHandlers(): void {
+    this.messagingSystem.registerActionHandler(
+      QueuedRequestControllerActionTypes.enqueueRequest,
+      this.enqueueRequest.bind(this),
+    );
+  }
+
+  length() {
```

Comment:
If `count` is stored in a public property, why do we need a method to access the property? It seems like we might want one way to do this.

Reply:
good call. I think the intent was to have the count be private. Will update accordingly, thanks!